### PR TITLE
Store typed version dependencies for workfiles

### PR DESCRIPTION
**Pull request by davidlatwe**
This PR is the first step for #2092 .

There are two types of links that will be created in this PR:
1. scene loaded versions -> workfile (reference type link)
2. workfile -> publishing instances (generative type link)

For linking specific loaded versions as direct dependencies to the publishing instance (e.g. specific rig to a pointcache), that would require a few more host specific PRs.

And for those future implementations, any active Pyblish plugin that sets/extends list of dependencies version ids into `instance.data['inputVersions']` will be handled by the new plugin `IntegrateInputLinks` in this PR.

## TODO
* [ ] Visulize those workfile centric links in Workfile App. (displaying what version of subsets has been used in scene)

### CODE REVIEW

**Comment by tokejepsen** [comment]
I know this is specific to versions but for assets we are using `inputs`; https://github.com/pypeclub/OpenPype/blob/develop/openpype/plugins/publish/extract_hierarchy_avalon.py#L59

**Comment by davidlatwe** [comment]
No worries, we have that covered. We use `inputLinks` :D

**Comment by davidlatwe** [comment]
So I find the Workfile App isn't a good place for showing the dependencies that relates to the workfile, since the workfiles over there aren't the published ones so the link infomation could be easily outdated.

How about this, a simple widget that can be embedded in the Loader or elsewhere ?

image

The above snapshot is where I am at the moment. As you can see, the list could be really long if we list every links in there. I am thinking maybe we could just displaying asset + family and the count ?

And if more detailed info is required, right click to show a context menu for further action, the action would maybe depends on Apps. For example, the Loader App could take those input/output version ids from the dependency widget and only show all those versions in the subset view.

**Comment by mkolar** [comment]
Task linked: [OP-1707 publish incoming and outgoing links](https://app.clickup.com/t/6658547/OP-1707)

File: openpype/plugins/publish/collect_scene_loaded_versions.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        "nuke",
+        "photoshop",
+        "resolve",
+        "tvpaint"
+    ]
+
+    def process(self, context):
+        from avalon import api, io
+
+        current_file = context.data.get("currentFile")
```
Comment (iLLiCiTiT) [inline openpype/plugins/publish/collect_scene_loaded_versions.py:27]: I don't think this is required...

File: openpype/plugins/publish/integrate_inputlinks.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        Args:
+            link_type (str): Type of link, one of 'reference' or 'generative'
+            input_id (str or ObjectId): Document Id of input version
+            version_doc (dict): The version document that takes the input
+
+        Returns:
+            None
+
+        """
+        from collections import OrderedDict
```
Comment (iLLiCiTiT) [inline openpype/plugins/publish/integrate_inputlinks.py:97]: I think you can put these imports to the top of file

**Comment by antirotor** [comment]
As I've tested it works really well. I am wondering if we should add option to open workfile in Dependency links in another process. So if you find what workfile was used to publish subset you are interested in, you could double-click on it and new Maya (for example) in the right context will fire up with that workfile opened?

File: openpype/plugins/publish/collect_scene_loaded_versions.py (hunk 1 above)
Comment (davidlatwe) [reply openpype/plugins/publish/collect_scene_loaded_versions.py:27]: Oh yeah, you are right. The code below and the other plugin still could work as expect even without this check.

File: openpype/plugins/publish/integrate_inputlinks.py (hunk 2 above)
Comment (davidlatwe) [reply openpype/plugins/publish/integrate_inputlinks.py:97]: This is kind of like a quirks of mine when I writing pyblish plugins, always afraid of modules getting garbage collected when previous plugin that also used the same modules has been processed.

I'll move them to the top and give it a go.

**Comment by davidlatwe** [comment]
> I am wondering if we should add option to open workfile...

Hmmm, or open up Workfile App and select the right workfile automatically ?

And how about the tiny widget that embedded in the Loader (image above) ? Any feedback on that ?

**Review by mkolar — APPROVED** [review]
This is working really well for the first iteration. I would actually leave it as is and work on using those links and better GUIs in further PRs. This is about publishing workfile links and that is being done very well.

**Comment by davidlatwe** [comment]
Sorry, forgot to fix what @iLLiCiTiT  has mentioned in his review, now should be ready !

**Comment by mkolar** [comment]
Happy to merge

**Pull request merged**